### PR TITLE
build: replace mocha, sinon and chai with jest

### PR DESCRIPTION
**Pull request by rmuchall**
Hi @vlapo ,

As discussed here is the first of two PRs.
This one contains Jest only. If/when it is accepted I'll do the eslint PR :)

Please let me know if you have any comments/concerns.
Thanks,
Rob

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        "@typescript-eslint/no-non-null-assertion": "off",
+        "@typescript-eslint/no-unused-vars": "off",
+        "@typescript-eslint/no-explicit-any": "off",
+        "@typescript-eslint/member-delimiter-style": "off",
+        "@typescript-eslint/member-ordering": "error",
+        "@typescript-eslint/unbound-method": ["error", {
+            "ignoreStatic": true
+        }]
+    }
+};
```
Comment (vlapo) [inline .eslintrc.js:33]: Lets do this in another PR :) Also applies to `tslint.json`

File: gulpfile.ts [hunk 2]
```diff
@@ -72,14 +68,10 @@ export class Gulpfile {
      */
     @Task()
     compileTests() {
```
Comment (vlapo) [inline gulpfile.ts:70]: `compileTests` should be `runTests`

File: gulpfile.ts [hunk 3]
```diff
@@ -241,55 +233,12 @@ export class Gulpfile {
...
-            .pipe(gulp.dest("./coverage"));
-    }
-
     /**
      * Compiles the code and runs tests.
      */
     @SequenceTask()
     tests() {
-        return ["clean", "compileTests", "tslint", "coveragePre", "coveragePost", "coverageRemap"];
+        return ["clean", "compileTests"];
```
Comment (vlapo) [inline gulpfile.ts:241]: Please do not remove `tslint` task from this sequence. We should force contributors to write tslint/eslint compliance code. Also I think it should be run before tests.

File: test/functional/custom-decorators.spec.ts [hunk 4]
```diff
@@ -1,261 +1,241 @@
...
-        }
-
-        class MyClass {
-            firstName: string;
-
-            @IsShortenThan("firstName", {
-                message: "$property must be shorter then $constraint1. Given value: $value"
-            })
-            lastName: string;
+describe("decorator with separate validation constraint class", function () {
```
Comment (vlapo) [inline test/functional/custom-decorators.spec.ts:175]: Please use everywhere arrow functions `() => {` :)

File: test/functional/validation-options.spec.ts [hunk 5]
```diff
@@ -20,1043 +12,1034 @@ const validator = new Validator();
 // Specifications: common decorators
 // -------------------------------------------------------------------------
 
-describe("validation options", function() {
```
Comment (vlapo) [inline test/functional/validation-options.spec.ts:23]: Why did you remove `describe`? I saw you remove root describe in some specs (but not at all of them). I would love to keep root describe.

File: test/functional/custom-decorators.spec.ts [hunk 6]
```diff
@@ -1,261 +1,241 @@
...
-
-// -------------------------------------------------------------------------
-// Setup
-// -------------------------------------------------------------------------
-
 const validator = new Validator();
 
-// -------------------------------------------------------------------------
-// Specifications: common decorators
-// -------------------------------------------------------------------------
```
Comment (vlapo) [inline test/functional/custom-decorators.spec.ts:24]: Please remove or keep these comments in all spec files :)

File: .eslintrc.js (hunk 1 above)
Comment (rmuchall) [reply .eslintrc.js:33]: My bad, I included this file by mistake.

File: gulpfile.ts (hunk 2 above)
Comment (rmuchall) [reply gulpfile.ts:70]: This is not a change that I made.

File: gulpfile.ts (hunk 3 above)
Comment (rmuchall) [reply gulpfile.ts:241]: I removed tslint because it appeared to not be linting any files.

File: test/functional/custom-decorators.spec.ts (hunk 4 above)
Comment (rmuchall) [reply test/functional/custom-decorators.spec.ts:175]: Yes, my mistake. I thought I had changed them all but missed this one.

File: test/functional/validation-options.spec.ts (hunk 5 above)
Comment (rmuchall) [reply test/functional/validation-options.spec.ts:23]: Jest does not require root describe blocks unless you wish to include before/after setup for all tests within the block. As it was superfluous I removed it.

File: test/functional/custom-decorators.spec.ts (hunk 6 above)
Comment (rmuchall) [reply test/functional/custom-decorators.spec.ts:24]: I will leave this one for you to fix ;)

**Comment by rmuchall** [comment]
> I thought about promises. Maybe we should unite code style (at least in tests) and use callback style or async/await

I attempted to adopt a consistent style wherever possible through all the tests. Most tests I refactored to return a promise. Those that don't were already configured to use a helper function and it was quicker just to modify those to use expect();

**Comment by rmuchall** [comment]
Thanks @vlapo :)
Thank you for your time and your patience.
I will do the eslint PR now.

RE: prettier
Prettier is not something that I use in my projects but I can try to install/configure it for class-validator if you wish it.

**Comment by vlapo** [comment]
> RE: prettier
> Prettier is not something that I use in my projects but I can try to install/configure it for class-validator if you wish it.

Lets start with eslint. IMHO I do not like prettier format :-D maybe eslint is enough for us

**Comment by rmuchall** [comment]
Yes, I feel the same as you. I do not like the prettier format either :)
